static/js/pages/CourseSearchPage.js

```diff
@@ -112,415 +63,442 @@ export type CourseSearchParams = {
-    const { incremental } = params
-    if (!incremental) {
-      clearSearch()
-      from = 0
-    }
-    this.setState({ from, incremental })
+  const runSearch = useCallback(
+    async (text, activeFacets, incremental = false) => {
+      let nextFrom = from + SETTINGS.search_page_size
 
-    if (emptyOrNil(activeFacets.type)) {
-      activeFacets.type = LR_TYPE_ALL
-    } else {
-      if (activeFacets.type.includes(LR_TYPE_PODCAST)) {
-        activeFacets.type.push(LR_TYPE_PODCAST_EPISODE)
+      if (!incremental) {
+        clearSearchCB()
+        nextFrom = 0
       }
+      setFrom(nextFrom)
+      setIncremental(incremental)
 
-      if (activeFacets.type.includes(LR_TYPE_USERLIST)) {
-        activeFacets.type.push(LR_TYPE_LEARNINGPATH)
-      }
-    }
+      const searchFacets = R.clone(activeFacets)
 
-    await runSearch({
-      channelName: null,
-      text,
-      type:        activeFacets.type,
-      // $FlowFixMe: type facet wont be undefined here
-      facets:      new Map(Object.entries(activeFacets)),
-      from,
-      size:        SETTINGS.search_page_size
-    })
-  }
+      if (emptyOrNil(searchFacets.type)) {
+        searchFacets.type = LR_TYPE_ALL
+      } else {
+        if (searchFacets.type.includes(LR_TYPE_PODCAST)) {
+          searchFacets.type.push(LR_TYPE_PODCAST_EPISODE)
+        }
 
-  debouncedRunSearch = _.debounce(this.runSearch, 500)
+        if (searchFacets.type.includes(LR_TYPE_USERLIST)) {
+          searchFacets.type.push(LR_TYPE_LEARNINGPATH)
+        }
+      }
 
-  setSearchUI = (searchResultLayout: string) => {
-    this.setState({
-      searchResultLayout
-    })
-  }
+      await dispatch(
+        actions.search.post({
+          channelName: null,
+          text,
+          type:        searchFacets.type,
+          facets:      new Map(Object.entries(searchFacets)),
+          from:        nextFrom,
+          size:        SETTINGS.search_page_size
+        })
+      )
 
-  toggleFacet = async (name: string, value: string, isEnabled: boolean) => {
-    const { location } = this.props
-    const { activeFacets } = deserializeSearchParams(location)
+      // search is updated, now echo params to URL bar
+      history.replace({
+        pathname: location.pathname,
+        search:   serializeSearchParams({
+          text,
+          activeFacets
+        })
+      })
+    },
+    [from, location, history, setFrom, setIncremental, clearSearchCB, dispatch]
+  )
 
-    if (isEnabled) {
-      activeFacets[name] = _.union(activeFacets[name] || [], [value])
-    } else {
-      activeFacets[name] = _.without(activeFacets[name] || [], value)
-    }
+  const loadMore = useCallback(
+    () => {
+      if (!loaded) {
+        // this function will be triggered repeatedly by <InfiniteScroll />, filter it to just once at a time
+        return
+      }
 
-    this.updateSearchState({
-      activeFacets
-    })
-  }
+      runSearch(text, activeFacets, true)
```

Comment:
Should there be an `await` here?

Reply:
you know, I just copied over the `clearSearchCB` from the class component, idk why that was implemented as an async function because there is nothing async about it.